dartagnan/src/main/java/com/dat3m/dartagnan/Dartagnan.java

```diff
@@ -223,13 +197,119 @@ public static void main(String[] args) throws Exception {
         }
     }
 
+	private static void generateWitnessIfAble(VerificationTask task, ProverEnvironment prover, ModelChecker modelChecker) {
+		// ------------------ Generate Witness, if possible ------------------
+		final EnumSet<Property> properties = task.getProperty();
+		if (modelChecker.hasModel() && properties.contains(PROGRAM_SPEC)) {
+			try {
+				WitnessBuilder w = WitnessBuilder.of(modelChecker.getEncodingContext(), prover, modelChecker.getResult());
+				if (w.canBeBuilt()) {
+					//  We can only write witnesses if the path to the original C file was given.
+					w.build().write();
+				}
+			} catch (InvalidConfigurationException e) {
+				logger.warn(e.getMessage());
+			}
+		}
+	}
+
+	public static String generateResultSummary(VerificationTask task, ProverEnvironment prover, ModelChecker modelChecker) throws SolverException {
+		// ----------------- Generate output of verification result -----------------
+		final Program p = task.getProgram();
+		final Result result = modelChecker.getResult();
+		final EncodingContext encCtx = modelChecker.getEncodingContext();
+		final Model model = modelChecker.hasModel() ? prover.getModel() : null;
+		final boolean hasViolations = result == FAIL && (model != null);
+		final boolean hasPositiveWitnesses = result == PASS && (model != null);
+
+		StringBuilder summary = new StringBuilder();
+
+		if (p.getFormat().equals(SourceLanguage.BOOGIE)) {
+			if (hasViolations) {
+				printWarningIfThreadStartFailed(p, encCtx, prover);
+				if(FALSE.equals(model.evaluate(PROGRAM_SPEC.getSMTVariable(encCtx)))) {
+					summary.append("Program specification violation found").append("\n");
+					summary.append("User assertion: ").append(p.getSpecification().toStringWithType()).append("\n");
+				}
+				if(FALSE.equals(model.evaluate(LIVENESS.getSMTVariable(encCtx)))) {
+					summary.append("Liveness violation found").append("\n");
+				}
+				final List<Axiom> violatedCATSpecs = task.getMemoryModel().getAxioms().stream()
+						.filter(Axiom::isFlagged)
+						.filter(ax -> FALSE.equals(model.evaluate(CAT_SPEC.getSMTVariable(ax, encCtx))))
+						.collect(Collectors.toList());
+				if (!violatedCATSpecs.isEmpty()) {
+					summary.append("CAT specification violation found").append("\n");
+					for (Axiom violatedAx : violatedCATSpecs) {
+						summary.append("Flag ")
+								.append(Optional.ofNullable(violatedAx.getName()).orElse(violatedAx.getNameOrTerm()))
+								.append("\n");
+					}
+				}
+			} else if (hasPositiveWitnesses) {
+				if(TRUE.equals(model.evaluate(PROGRAM_SPEC.getSMTVariable(encCtx)))) {
+					// The check above is just a sanity check: the program spec has to be true
+					// because it is the only property that got encoded.
+					summary.append("Program specification witness found.").append("\n");
```

Comment:
Shouldn't we add
```
summary.append("User assertion: ").append(p.getSpecification().toStringWithType()).append("\n");
```
for consistency with the violation case?

Reply:
Yes, seems I forgot it.